clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -983,6 +987,91 @@ private Node atLeastThreeLeastLoadedNode(long now, List<Node> nodes) {
         return chosenNode;
     }
 
+    private Node randomLeastLoadedNode(long now, List<Node> nodes) {
+        int inflight = Integer.MAX_VALUE;
+
+        Node foundConnecting = null;
+        Node foundCanConnect = null;
+        Node foundReady = null;
+
+        int offset = this.randOffset.nextInt(nodes.size());
```

Comment:
There is a small chance that, depending on how good the random number generation is, you end up not evenly distributing requests amongst eligible brokers with this strategy.

It would probably be best if you could maintain a list of brokers to select which is randomly shuffled.

```
if (candidateBrokerList is empty) {
  add broker id of known `Node` objects to candidateBrokerList
  shuffle(candidateBrokerList)
}

do {
candidateBrokerId = candidateBrokerList.pop()
} (while canSendREquest(node));

return lookupNode(candidateBrokerId)
```

That way each broker in the cluster will always visit peers they know about round-robin, ensuring they ask every broker at least once, but they all ask in a different order.

When this code learns about new `Node`s it hasn't seen before, they can just be added to the end of the candidateBrokerList queue, since it will be randomly shuffled once it is empty.

Reply:
on top of maintaining the added datastructure (list of known brokers in random order plus index) and the risks involved with failing to maintain that "in sync", the resulting behavior (random order round robin where all brokers are guaranteed to be selected before the same broker is ever re-selected) isnt really "random", and actually would completely maximize the number of sockets established (even faster than a pure random).

i'd rather assume the distribution of random is good enough for govt work ? :-)

Reply:
As long as you don't think it would be an issue then sure I think it's fine to just go with what you have.